### ISSUE TITLE
关于实现和paper的区别，请指点一下

### ISSUE THREAD

**Issue by Yitian-Li** [huanghoujing/person-reid-triplet-loss-baseline#36]
大佬你好，关于market1501数据集，paper中TriNet的mAP大概是69.1%，我在stride1的情况下训练出来大概是71%，差2%感觉有点多，自己也仔细的对照了一下区别。

下面的3个问题有点啰嗦，其实我就是想问一下：如果想达到69.1%（而不是71%）的mAP，需要调哪些args？如果你时间比较紧，可以忽略下面的问题，回答这个就足够了。十分期待你的回答，谢谢！

1.训练集共有751个人，按照P=32，算下来就是751/32=24个iteration=1epoch，但是为什么paper中他搞了25k个iteration啊，算下来就是1000+个epoch左右了，这和300个epoch比起来也太夸张了吧。是有什么地方我没算对么？
2.数据增强。我看原文中是"four corner crops and one center crop, as well as a horizontally flipped copy of each"，代码中train_set是通过create_dataset函数创建的，但是create_dataset函数中没有关于数据增强的部分，请问数据增强是在什么地方做的？
3.我看model创建的时候pretrain默认是True，所以不管用trainval还是train，初始模型都是ImageNet上的ResNet50的预训练模型，对吧?

**Comment by Yitian-Li** [comment]
噢，不好意思，我的训练是stride2，写错了。

我用的命令：
python script/experiment/train.py -d '(0,)' 
--only_test false 
--dataset market1501 
--last_conv_stride 2 
--normalize_feature false 
--trainset_part trainval 
--exp_dir TRAIN-S2-gpu0 
--steps_per_log 50 
--epochs_per_val 10

结果：
=========> Test on dataset: market1501 <=========

Extracting feature...
1000/1000 batches done, +1.73s, total 87.83s
Done, 88.01s
Computing distance...
Done, 0.83s
Computing scores...
User Warning: Version 0.18.1 is required for package scikit-learn, your current version is 0.20.4. As a result, the mAP score may not be t uninstall scikit-learn` and then `pip install scikit-learn==0.18.1`
Done, 15.43s
Single Query:                 [mAP: 71.78%], [cmc1: 86.73%], [cmc5: 94.83%], [cmc10: 96.97%]
Multi Query, Computing distance...
Done, 0.74s
Multi Query, Computing scores...
User Warning: Version 0.18.1 is required for package scikit-learn, your current version is 0.20.4. As a result, the mAP score may not be t uninstall scikit-learn` and then `pip install scikit-learn==0.18.1`
Done, 14.95s
Multi Query:                  [mAP: 79.54%], [cmc1: 91.42%], [cmc5: 96.94%], [cmc10: 98.16%]
Re-ranking distance...
./tri_loss/utils/distance.py:35: RuntimeWarning: invalid value encountered in sqrt
  dist = np.sqrt(squared_dist)
Done, 91.14s

**Comment by huanghoujing** [comment]
您好，感谢关注！
这个代码和文章有一些细节不同，而且有TensorFlow和Pytorch的差异，可能复现不出来文章的分数。
1. 文章按照iteration数来算的话，那确实相当于训了很多epoch
2. 数据增强在[这里](https://github.com/huanghoujing/person-reid-triplet-loss-baseline/blob/master/tri_loss/dataset/PreProcessImage.py#L76)，这份代码只用了镜像对称
3. 是的

**Comment by Yitian-Li** [comment]
> 您好，感谢关注！
> 这个代码和文章有一些细节不同，而且有TensorFlow和Pytorch的差异，可能复现不出来文章的分数。
> 
> 1. 文章按照iteration数来算的话，那确实相当于训了很多epoch
> 2. 数据增强在[这里](https://github.com/huanghoujing/person-reid-triplet-loss-baseline/blob/master/tri_loss/dataset/PreProcessImage.py#L76)，这份代码只用了镜像对称
> 3. 是的

**Comment by Yitian-Li** [comment]
那么请问除了stride不一样之外，还有哪些差别呢

**Comment by huanghoujing** [comment]
1. 训练迭代次数
2. PK采样的时候，一个batch内有多少人，每个人多少张图片
3. 网络backbone后面有一个、或两个全连接层

其它地方的差异我也已经记不清楚了，得翻出In defense那篇论文仔细找找才知道了。

**Comment by Yitian-Li** [comment]
非常感谢！
